tests/test_markdown_plugin.py

```diff
@@ -0,0 +1,187 @@
+IS_LIST_ITEM_CASES.extend(
+    [
+        IsListItemCase(
+            id=f"raw prexix {prefix} no space",
+            line=f"{prefix}{' ' if space else ''}",
+            expected=space,
+        )
+        for prefix, space in itertools.product(
+            ["1.", "1)", "#.", "#)", "-", "*", "+"], [True, False]
+        )
+    ]
+)
+
+# test numbered list with whitespace following and preceding
+IS_LIST_ITEM_CASES.extend(
+    [
+        IsListItemCase(
+            id=f"numbered {preceding=} {following=} space",
+            line=f"{' ' * preceding}{i}{sep}{' ' * following} item 1",
+            expected=True,
+        )
+        for i, sep, preceding, following in itertools.product(
+            ("7", "#"), (".", ")"), range(11), range(11)
+        )
+    ]
+)
+
+# test with whitespace following and preceding
+IS_LIST_ITEM_CASES.extend(
+    [
+        IsListItemCase(
+            id=f"bullet {preceding=} {following=} space",
+            line=f"{' ' * preceding}{bullet} {' ' * following} item 1",
+            expected=True,
+        )
+        for bullet, preceding, following in itertools.product(("-", "*", "+"), range(11), range(11))
+    ]
+)
+
+
+@pytest.mark.parametrize(
+    "line, expected, raises",
+    [
+        pytest.param(
+            case.line,
+            case.expected,
+            pytest.raises(case.raises) if case.raises else does_not_raise(),
+            marks=case.marks,
+            id=case.id,
+        )
+        for case in IS_LIST_ITEM_CASES
+    ],
+)
+def test_is_list(line: str, expected: bool, raises):
+    with raises:
+        assert markdown._is_list_item(line) == expected
+
+
+@pytest.mark.parametrize(
+    "li",
+    [
+        "-",
+        "1.",
+        "1. item 1",
+        "1) item 1",
+        "#) item 1",
+        "- item 1",
+        "* item 1",
+        "+ item 1",
+        "+ +++++ weird",
+        "1) ))))) still weird",
+        "- [ ] unchecked task",
+        "- [X] checked task",
+    ],
+)
+def test_filetype_switching(li: str, filetab, tmp_path):
+    assert filetab.settings.get("filetype_name", object) == "Python"
+
+    filetab.textwidget.insert("1.0", li)
+    filetab.textwidget.event_generate("<Tab>")
+    filetab.update()
+
+    assert (
+        filetab.textwidget.get("1.0", "insert") == li
+    ), "should not effect list items unless using markdown filetype"
+    filetab.textwidget.event_generate("<Escape>")  # close the autocomplete
+
+    # switch to Markdown filetype format
+    filetab.save_as(tmp_path / "asdf.md")
+    assert filetab.settings.get("filetype_name", object) == "Markdown"
+
+    filetab.textwidget.event_generate("<Tab>")
+    filetab.update()
+    # no change to text, should open autocomplete menu
+    assert filetab.textwidget.get("1.0", "insert") == li
+    filetab.textwidget.event_generate("<Escape>")  # close the autocomplete
+
+    # add a space
+    filetab.textwidget.insert("insert", " ")
+    filetab.textwidget.event_generate("<Tab>")
+    filetab.update()
```

Comment:
I find that you often need to `.update()` before `event_generate()`, so changing the order might help.

Reply:
Should we consider adding some helper utilities for these sorts of things?

Reply:
I don't think I want to hide the `.update()` + `.event_generate()` combination, since it's just two lines and many people already know what `.update()` does. It should be documented much better though (#1308)